Add appendix (#11)
jhoyla/draft-jhoyla-req-mtls-flag: draft-jhoyla-req-mtls-flag.md

--- back

# Appendix

## Alternative Approaches

### Concealed Auth

> Concealed Auth {{?RFC9729}} is a draft that allows a client to attach a header

Reviewer: Not a draft anymore ;)
```suggestion
Concealed Auth {{?RFC9729}}1 allows a client to attach a header
```

at the HTTP layer that cryptographically binds a public key to the underlying
TLS connection. This gives us a similar authentication property to using
`request_client_auth`, and is also client initiated. This solution is applicable

> in some contexts, but in the case of authenticating bots is difficult to

Reviewer: ```suggestion
in some contexts, but in the case of authenticating bots may be difficult to
```

Reviewer: Also, why would this be hard to implement specifically for the bots use case and not others?

implement.

The way many HTTP libraries work is that they dispatch requests to a connection
pool. Once the request arrives at the pool it is sent on a chosen connection
driven by internal logic. This, unfortunately, makes it difficult to know which
connection a request will be sent on, and thus to produce the appropriate
signature.

> 
> It is possible to avoid this issue, for example by not using connection pooling,
> or by sending a special "pre-request" on each connection that's added to a pool,
> however these solutions tend to be bespoke per library and bring a number of
> their own issues (e.g. handling servers that terminate connections after a
> single request, or handling resumption).

Reviewer: We should mention the possibility of modifying the pooling code.
```suggestion

It is possible to avoid this issue, for example by not using connection pooling, or modifying
the implementation so that the first request on each new connection includes the binding.
If modifying the implementation is not possible, then there are some ad-hoc options, such as
sending a special "pre-request" on each connection that's added to a pool,
however these solutions tend to be bespoke per library and bring a number of
their own issues (e.g. handling servers that terminate connections after a
single request, or handling resumption).
```

### Post-handshake authentication (PHA)

Post-handshake authentication is part of the core TLS 1.3 RFC, although it is
not widely supported. Its use incurs an extra round-trip, and would need
negotiation at the application layer. This just trades one negotiation for
another, whilst making the protocol more expensive to run. Further there are

> some reservations about its security guarantees.

Reviewer: Add a reference for these reservations.

### ALPN

One proposal is for the client to send new ALPN extensions, e.g. `http2+bot`.
This is functionally equivalent to this draft, but with a less efficient

> encoding.

Reviewer: Say why this is functionally equivalent. We might need to flesh out the proposal a bit more.

> ### HTTP Signatures

Reviewer: I would put this section first since this draft shares the same exact motivation as ours. I would also mention the shared motivation explicitly.

HTTP Signatures {{?I-D.draft-meunier-web-bot-auth-architecture}} describes a
mechanism for implementing bearer tokens at the HTTP layer that authenticate
requests. This mechanism suffers from a number of drawbacks, notably that a new
signature must be created and verified for every request (rather than one per
connection), and second that, because the tokens are not bound to the underlying

> TLS connection they can be stolen and replayed.

Reviewer: Mention the CDN use case specifically.

--- back

# Appendix

## Alternative Approaches

### HTTP Signatures

HTTP Signatures {{?I-D.draft-meunier-web-bot-auth-architecture}} describes a
mechanism for implementing bearer tokens at the HTTP layer that authenticate

> requests. This mechanism suffers from a number of drawbacks, notably that a new

Author: A bearer token is anything that simple possession of grants access, i.e. they are always transferable. It doesn't need to involve a shared secret, it can be, for example a signed blob.

signature must be created and verified for every request (rather than one per
connection), and second that, because the tokens are not bound to the underlying
TLS connection they can be stolen and replayed. This is especially problematic

> in the multi-CDN use case. If a bot creates a token for a realm that is provided

Reviewer: ```suggestion
in the multi-CDN ("Content Distribution Network") use case. If a bot creates a token for a realm that is provided
```

by more than one CDN then the CDN that receives that token can use it to attack
any of the other CDNs. This means that all CDNs are required to trust each
other.

### Concealed Auth

Concealed Auth {{?RFC9729}} allows a client to attach a header at the HTTP
layer that cryptographically binds a public key to the underlying TLS
connection. This gives us a similar authentication property to using
`request_client_auth`, and is also client initiated. This solution is

> applicable in some contexts, but in the case of authenticating bots may be
> difficult to implement.

Reviewer: I don't see how the next paragraph is about bots specifically: see  but in the case of authenticating bots may be
difficult to implement.

```suggestion
applicable in some contexts, but may be
difficult to implement.
```

The way many HTTP libraries work is that they dispatch requests to a connection
pool. Once the request arrives at the pool it is sent on a chosen connection
driven by internal logic. This, unfortunately, makes it difficult to know which
connection a request will be sent on, and thus to produce the appropriate
signature.

It is possible to avoid this issue, for example by not using connection
pooling, or by modifying the implementation to add the header on the first
request of each connection. Where modifying the implementation is not possible
various ad-hoc solutions can be used such as sending a special "pre-request" on

> each connection that's added to a pool, however these solutions tend to be

Reviewer: ```suggestion
each connection that's added to a pool. However these solutions tend to be
```

bespoke per library and bring a number of their own issues (e.g. handling
servers that terminate connections after a single request, or handling
resumption).

> ### Post-handshake authentication (PHA)

Reviewer: nit: consistent title capitalization (or let me know if you're following a convention I'm aware of)
```suggestion
### Post-handshake Authentication (PHA)
```
